pkg/rotation/reconciler.go

```diff
@@ -0,0 +1,346 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"sigs.k8s.io/secrets-store-csi-driver/pkg/k8s"
+
+	"sigs.k8s.io/secrets-store-csi-driver/pkg/util/fileutil"
+	"sigs.k8s.io/secrets-store-csi-driver/pkg/util/secretutil"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	secretsstore "sigs.k8s.io/secrets-store-csi-driver/pkg/secrets-store"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/secrets-store-csi-driver/apis/v1alpha1"
+	secretsStoreClient "sigs.k8s.io/secrets-store-csi-driver/pkg/client/clientset/versioned"
+
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	permission os.FileMode = 0644
+)
+
+// Reconciler reconciles and rotates contents in the pod
+// and Kubernetes secrets periodically
+type Reconciler struct {
+	store                k8s.Store
+	ctrlReaderClient     client.Reader
+	ctrlWriterClient     client.Writer
+	providerVolumePath   string
+	scheme               *runtime.Scheme
+	rotationPollInterval time.Duration
+	providerClients      map[string]*secretsstore.CSIProviderClient
+}
+
+// NewReconciler returns a new reconciler for rotation
+func NewReconciler(s *runtime.Scheme, providerVolumePath, nodeName string, rotationPollInterval time.Duration) (*Reconciler, error) {
+	config, err := buildConfig()
+	if err != nil {
+		return nil, err
+	}
+	kubeClient := kubernetes.NewForConfigOrDie(config)
+	crdClient := secretsStoreClient.NewForConfigOrDie(config)
+	c, err := client.New(config, client.Options{Scheme: s, Mapper: nil})
+	if err != nil {
+		return nil, err
+	}
+	store, err := k8s.New(kubeClient, crdClient, nodeName, 60*time.Second)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Reconciler{
+		store:                store,
+		ctrlReaderClient:     c,
+		ctrlWriterClient:     c,
+		scheme:               s,
+		providerVolumePath:   providerVolumePath,
+		rotationPollInterval: rotationPollInterval,
+		providerClients:      make(map[string]*secretsstore.CSIProviderClient),
+	}, nil
+}
+
+// Run starts the rotation reconciler
+func (r *Reconciler) Run(stopCh <-chan struct{}) {
+	log.Infof("starting rotation reconciler with poll interval: %s", r.rotationPollInterval)
+
+	ticker := time.NewTicker(r.rotationPollInterval)
+	defer ticker.Stop()
+
+	if err := r.store.Run(stopCh); err != nil {
+		log.Fatalf("failed to run informers for rotation reconciler, err: %+v", err)
+	}
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case <-ticker.C:
+			log.Infof("rotation reconcile started")
+			// The spc pod status informer is configured to do a filtered list watch of spc pod statuses
+			// labeled for the same node as the driver. LIST will only return the filtered results.
+			spcpsList, err := r.store.ListSecretProviderClassPodStatus()
+			if err != nil {
+				log.Errorf("[rotation] failed to list secret provider class pod status for node, err: %+v", err)
+				continue
+			}
+			for _, spcps := range spcpsList {
```

Comment:
What happens if this loop takes longer than the rotation poll interval?

Reply:
I've updated this part to add the objects to the queue which are then processed by the worker. But even without the change, the `<-ticker.C` will be blocked until it's read from the channel. So if the loop takes longer than the rotation poll interval, the last tick will still be in the channel and until it's read from the channel, subsequent ticks will be dropped.

Reply:
Thanks for adding the queue! maybe as a followup, let's add duration for each run to our metrics. Would be useful for debugging and performance tests.